pkg/shp/bundle/bundle.go

```diff
@@ -0,0 +1,245 @@
+
+			case update, ok := <-updates:
+				if !ok {
+					return
+				}
+
+				if progress == nil {
+					progress = progressbar.NewOptions(int(update.Total),
+						progressbar.OptionSetWriter(io.ErrOut),
+						progressbar.OptionEnableColorCodes(true),
+						progressbar.OptionShowBytes(true),
+						progressbar.OptionSetWidth(15),
+						progressbar.OptionSetPredictTime(false),
+						progressbar.OptionSetDescription("Uploading local source..."),
+						progressbar.OptionSetTheme(progressbar.Theme{
+							Saucer:        "[green]=[reset]",
+							SaucerHead:    "[green]>[reset]",
+							SaucerPadding: " ",
+							BarStart:      "[",
+							BarEnd:        "]"}),
+						progressbar.OptionClearOnFinish(),
+					)
+					defer progress.Close()
+				}
+
+				progress.ChangeMax64(update.Total)
+				_ = progress.Set64(update.Complete)
+			}
+		}
+	}()
+
+	fmt.Fprintf(io.Out, "Bundling %q as %q ...\n", localDirectory, targetImage)
+	digest, err := buildbundle.PackAndPush(
+		tag,
+		localDirectory,
+		remote.WithContext(ctx),
+		remote.WithAuth(auth),
+		remote.WithProgress(updates),
+	)
+
+	done <- struct{}{}
+	return digest, err
+}
+
+func Prune(ctx context.Context, io *genericclioptions.IOStreams, image string) error {
+	ref, err := name.ParseReference(image)
+	if err != nil {
+		return err
+	}
+
+	auth, err := authn.DefaultKeychain.Resolve(ref.Context())
+	if err != nil {
+		return err
+	}
+
+	// Deleting a tag, or a whole repo is not as straightforward as initially
+	// planned as DockerHub seems to restrict deleting a single tag for
+	// standard users. This might be subject to change, but as of September
+	// 2021 it is limited to the business tier. However, there is an API call
+	// to delete the whole repository. In case there is only one tag used in
+	// a repository, the effect is pretty much the same. For convenience, there
+	// is a provider switch to deal with images on DockerHub differently.
+	//
+	// DockerHub images:
+	// - In case the repository only has one tag, the repository is deleted.
+	// - If there are multiple tags, the tag to be deleted is overwritten
+	//   with an empty image (to remove the content, and save quota).
+	// - Edge case would be no tags in the repository, which is ignored.
+	//
+	// Other registries:
+	// Use standard spec delete API request to delete the provided tag.
+	//
+	switch ref.Context().RegistryStr() {
+	case "index.docker.io":
+		list, err := remote.List(ref.Context(), remote.WithContext(ctx), remote.WithAuth(auth))
+		if err != nil {
+			return err
+		}
+
+		switch len(list) {
+		case 0:
+			return nil
+
+		case 1:
+			authr, err := auth.Authorization()
+			if err != nil {
+				return err
+			}
+
+			token, err := dockerHubLogin(authr.Username, authr.Password)
+			if err != nil {
+				return err
+			}
+
+			return dockerHubRepoDelete(token, ref)
+
+		default:
+			fmt.Fprintf(io.ErrOut, "removing specific tags is currently not supported for images hosted on %s\n", ref.Context().RegistryStr())
```

Comment:
I think this needs to be modified as it's contradictory to the fact that one can remove a tag, as long as the container registry holds one single tag. I do understand that if multiple tags are present, then this msg will be triggered, but the msg should be different. What do you think?

Reply:
Working on rephrasing this one.